test/Microsoft.AspNetCore.Mvc.TagHelpers.Test/FormActionTagHelperTest.cs

```diff
@@ -0,0 +1,387 @@
+using System.Reflection;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Mvc.ModelBinding;
+using Microsoft.AspNetCore.Mvc.Routing;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using Microsoft.AspNetCore.Routing;
+using Moq;
+using Xunit;
+
+namespace Microsoft.AspNetCore.Mvc.TagHelpers
+{
+    public class FormActionTagHelperTest
+    {
+        [Fact]
+        public async Task ProcessAsync_GeneratesExpectedOutput()
+        {
+            // Arrange
+            var expectedTagName = "not-button-or-submit";
+            var metadataProvider = new TestModelMetadataProvider();
+
+            var tagHelperContext = new TagHelperContext(
+                allAttributes: new TagHelperAttributeList
+                {
+                    { "id", "my-id" },
+                    { "asp-route-name", "value" },
+                    { "asp-action", "index" },
+                    { "asp-controller", "home" },
+                },
+                items: new Dictionary<object, object>(),
+                uniqueId: "test");
+            var output = new TagHelperOutput(
+                expectedTagName,
+                attributes: new TagHelperAttributeList
+                {
+                    { "id", "my-id" },
+                },
+                getChildContentAsync: (useCachedResult, encoder) =>
+                {
+                    var tagHelperContent = new DefaultTagHelperContent();
+                    tagHelperContent.SetContent("Something Else");
+                    return Task.FromResult<TagHelperContent>(tagHelperContent);
+                });
+            output.Content.SetContent("Something");
+
+            var urlHelper = new Mock<IUrlHelper>();
+            urlHelper
+                .Setup(mock => mock.Action(It.IsAny<UrlActionContext>())).Returns("home/index").Verifiable();
+
+            var urlHelperFactory = new Mock<IUrlHelperFactory>();
+            urlHelperFactory
+                .Setup(f => f.GetUrlHelper(It.IsAny<ActionContext>()))
+                .Returns(urlHelper.Object);
+
+            var htmlGenerator = new TestableHtmlGenerator(metadataProvider, urlHelper.Object);
+            var viewContext = TestableHtmlGenerator.GetViewContext(
+                model: null,
+                htmlGenerator: htmlGenerator,
+                metadataProvider: metadataProvider);
+            var tagHelper = new FormActionTagHelper(urlHelperFactory.Object)
+            {
+                Action = "index",
+                Controller = "home",
+                RouteValues =
+                {
+                    {  "name", "value" },
+                },
+                ViewContext = viewContext,
+            };
+
+            // Act
+            await tagHelper.ProcessAsync(tagHelperContext, output);
+
+            // Assert
+            urlHelper.Verify();
+            Assert.Equal(2, output.Attributes.Count);
+            var attribute = Assert.Single(output.Attributes, attr => attr.Name.Equals("id"));
```

Comment:
Food for thought: If we were to move [this](https://github.com/aspnet/Razor/blob/dev/test/Microsoft.AspNetCore.Razor.Runtime.Test/Runtime/TagHelpers/CaseSensitiveTagHelperAttributeComparer.cs) to the test sources project we could use this in these various tests.